src/autora/experimentalist/random_.py

```diff
@@ -0,0 +1,341 @@
+        >>> random_pool(
+        ...     VariableCollection(independent_variables=[
+        ...         Variable(name="x", value_range=(-5, 5))
+        ... ]), random_state=1)["conditions"]
+                  x
+        0  0.118216
+        1  4.504637
+        2 -3.558404
+        3  4.486494
+        4 -1.881685
+
+
+
+        The allowed_values or value_range must be specified:
+        >>> random_pool(VariableCollection(independent_variables=[Variable(name="x")]))
+        Traceback (most recent call last):
+        ...
+        ValueError: allowed_values or [value_range and type==REAL] needs to be set...
+
+        With two independent variables, we get independent samples on both axes:
+        >>> random_pool(VariableCollection(independent_variables=[
+        ...         Variable(name="x1", allowed_values=range(1, 5)),
+        ...         Variable(name="x2", allowed_values=range(1, 500)),
+        ... ]), num_samples=10, replace=True, random_state=1)["conditions"]
+           x1   x2
+        0   2  434
+        1   3  212
+        2   4  137
+        3   4  414
+        4   1  129
+        5   1  205
+        6   4  322
+        7   4  275
+        8   1   43
+        9   2   14
+
+        If any of the variables have unspecified allowed_values, we get an error:
+        >>> random_pool(
+        ...     VariableCollection(independent_variables=[
+        ...         Variable(name="x1", allowed_values=[1, 2]),
+        ...         Variable(name="x2"),
+        ... ]))
+        Traceback (most recent call last):
+        ...
+        ValueError: allowed_values or [value_range and type==REAL] needs to be set...
+
+
+        We can specify arrays of allowed values:
+
+        >>> random_pool(
+        ...     VariableCollection(independent_variables=[
+        ...         Variable(name="x", allowed_values=np.linspace(-10, 10, 101)),
+        ...         Variable(name="y", allowed_values=[3, 4]),
+        ...         Variable(name="z", allowed_values=np.linspace(20, 30, 11)),
+        ... ]), random_state=1)["conditions"]
+             x  y     z
+        0 -0.6  3  29.0
+        1  0.2  4  24.0
+        2  5.2  4  23.0
+        3  9.0  3  29.0
+        4 -9.4  3  22.0
+
+
+    """
+    rng = np.random.default_rng(random_state)
+
+    raw_conditions = {}
+    for iv in variables.independent_variables:
+        if iv.allowed_values is not None:
+            raw_conditions[iv.name] = rng.choice(
+                iv.allowed_values, size=num_samples, replace=replace
+            )
+        elif (iv.value_range is not None) and (iv.type == ValueType.REAL):
+            raw_conditions[iv.name] = rng.uniform(*iv.value_range, size=num_samples)
+
+        else:
+            raise ValueError(
+                "allowed_values or [value_range and type==REAL] needs to be set for "
+                "%s" % (iv)
+            )
+
+    conditions = pd.DataFrame(raw_conditions)
+    return Result(conditions=conditions)
```

Comment:
Hmm... So the random function returns a data frame, but the grid function returns some kind of iterable? Or am I missing something? In any case, this should be consistent.

Reply:
You're right, makes no sense! grid is fixed in the next update.